cherab/core/model/plasma/total_radiated_power.pyx

```diff
@@ -60,22 +93,35 @@ cdef class TotalRadiatedPower(PlasmaModel):
 
         ni_upper = self._recom_species.distribution.density(point.x, point.y, point.z)
 
+        nhyd = 0
+        for hyd_species in self._hydrogen_species:
```

Comment:
Iterating over lists is optimised in Cython. I've checked the performance now and this particular loop has no impact on performance. However, since we now take into account three processes instead of two, performance has dropped by about 50%.
```python
        if self._plt_rate and ni > 0:  # excitation
            power_density += self._plt_rate.evaluate(ne, te) * ne * ni

        if self._prb_rate and ni_upper > 0:  # recombination + bremsstrahlung
            power_density += self._prb_rate.evaluate(ne, te) * ne * ni_upper

        if self._prc_rate and ni_upper > 0 and nhyd > 0:  # charge exchange
            power_density += self._prc_rate.evaluate(ne, te) * nhyd * ni_upper
```

Reply:
I wouldn't be concerned about speed here. If speed is an issue, there are other ways how to achieve speedup when calculating the total radiated power, by for example pre-calculating the powers and using `RadiationFunction` instead or using single bin spectra.

One improvement we could do in the future is to define separate models for individual components and the `TotalRadiatedPower` would become only a wrapper around them. This would allow control over which radiation is included and would allow users to do separate simulations showing the impact of the components.